File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiPicture.cpp
#include "StdAfx.h"
#include "UiPicture.h"
#include "Byo2Image.h"
#include "ImgImage.h"

#define ThisMethod($x) &UiPicture::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiPicture::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiPicture> c_obj;
	}

	void UiPicture::DefineControl()
	{
		AutoAddMethod( SetStretchMode );
		AutoAddMethod( GetStretchMode, WrapObjectGeneric );
		
		AutoAddMethod( SetImageFilter );
		AutoAddMethod( GetImageFilter, WrapObjectGeneric );

		AutoAddMethod( SetPicture );
		AutoAddMethod( SetImageData );
		AutoAddMethod( SetImage );

		AutoAddMethod( GetAnimationInfo, WrapObjectGeneric );
		AutoAddMethod( GetDuration, WrapObjectGeneric );

		AutoAddMethod( Play, WrapObjectGeneric );
		AutoAddMethod( Stop, WrapObjectGeneric );
		AutoAddMethod( IsPlaying, WrapObjectGeneric );

		AutoAddMethod( SetPlayPosition );
		AutoAddMethod( GetPlayPosition, WrapObjectGeneric );
		
		AutoAddMethod( SetPlayFrame );
		AutoAddMethod( GetPlayFrame, WrapObjectGeneric );
	}

	U1 UiPicture::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	UiPicture * UiPicture::SetImageData( ImgImage * img )
	{
		GetControlTyped().SetImageData( img->CloneImage() );
		return this;
	}

	UiPicture * UiPicture::SetImage( Byo2Image * img )
	{
		GetControlTyped().SetImage( img->CloneImage() );
		return this;
	}

	WrapData<Img::AnimationInfo> UiPicture::GetAnimationInfo() const
	{
		return GetControlTyped().GetAnimationInfo();
	}

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Visual/UiPicture.ts
import { AveLib } from "../../AveLib";
import { IVisual } from "../UiVisual";
import { WindowLike } from "../Control/UiWindow";
import { StretchMode } from "../UiCommon";
import { Byo2Image } from "../../Ave";
import { ResourceSource } from "../../Io/IoCommon";
import { DrawImageFilter } from "../UiPainter";
import { AnimationInfo, AveImage } from "../../Image";

export interface IPicture extends IVisual {
	new(window: WindowLike): IPicture;

	SetStretchMode(nMode: StretchMode): Picture;
	GetStretchMode(): StretchMode;

	SetImageFilter(nFilter: DrawImageFilter): Picture;
	GetImageFilter(): DrawImageFilter;

	SetPicture(rs: ResourceSource): Picture;
	SetImageData(img: AveImage): Picture;
	SetImage(img: Byo2Image): Picture;

	GetAnimationInfo(): AnimationInfo;
	GetDuration(): number;

	Play(): Picture;
	Stop(): Picture;
	IsPlaying(): boolean;

	SetPlayPosition(fTime: number): Picture;
	GetPlayPosition(): number;

	SetPlayFrame(nFrame: number): Picture;
	GetPlayFrame(): number;
}

export class Picture extends (AveLib.UiPicture as IPicture) { }
